### PR TITLE
Cirrus: Simplify use of cache in automation

### PR DESCRIPTION
**Pull request by cevich**
***Depends on:*** #13376 #14180

With the increasingly complex `.cirrus.yml` task relationships, build
cache wasn't always working as intended.  Recently, non-build tasks were
observed assuming authority over `gopath_cache`.  Ref.:
https://github.com/containers/podman/pull/13998#issuecomment-1108834538

Address this by an overall simplification using artifacts instead of
cache.  Using artifacts allows establishing concrete
authorship/authority over cached repo. content.  In this way, dependent tasks
may simply consume the artifact with `curl` instead of needing to use
complex algorithms.

Also, add support for new automation library version w/ `$DEBUG` fix.
ref: https://github.com/containers/automation_images/pull/128

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by cevich** [comment]
Note to me: This needs https://github.com/containers/automation_images/pull/128

**Comment by edsantiago** [comment]
BTW I'm not ignoring your review request, or actually I am, but only until CI passes and the churn stops. If you'd like me to look at it while it's still in turmoil, please lmk. Otherwise I'll wait for green CI.

**Comment by cevich** [comment]
Okay, I think this is ready (tests running as I write this) at least for review.  There may be some test failures & tweaking needed due to switching some tasks from VMs -> containers.  But overall this PR is doing what I intended, eeking out several more minutes of runtime savings and cleaning up the horrible/confusing caching mess in `.cirrus.yml`.

**Comment by TomSweeneyRedHat** [comment]
rootless tests is having trouble wit ssh.
Otherwise LGTM, but I'm not a SME by any means.

**Comment by cevich** [comment]
> rootless tests is having trouble wit ssh

Ya, not sure why that suddenly started happening.  I'm testing out a fix for it.

**Comment by cevich** [comment]
Force-pushing, rebased and added simplified rootless ssh setup.

**Comment by cevich** [comment]
[Argh! rootless unit is still borkt.](https://cirrus-ci.com/task/5468821833121792).  Thing is, "it works fine for me in a VM" (C).

Sigh...okay, I guess some more tinkering is needed here.

**Comment by cevich** [comment]
Oh! I found the problem, it's caused by trying to run the rootless unit test inside a container, where of course there is no sshd running and `localhost` means something different.  Okay, force-pushing with that fix and a split of the `$DEBUG` fix into a dedicated commit (in case "use containers" commit needs to ever be reverted).

**Comment by cevich** [comment]
The rootless unit problem is fixed, and I'm seeing other problems (queuing) due to swiching tasks to use containers.  This is unfortunate, since (absent task-queuing) the container-based tasks run a bit faster than VMs.  I'm going to [remove that commit](98ae246f3f0ed99396eeb748f8b7b373b1b501c3) and leave it up to a future effort.

File: Makefile
Comment (edsantiago) [inline Makefile:891]: I'm sorry, I can't allow this. It's incredibly dangerous: I've just spent a long time testing it, and have confirmed that the `unexport` carries over for the lifetime of the `make`. So this is fine for someone calling `make .install.bats`, or (shudder) in the current case where `.install.bats` is the last component of `install.tools`, but should anyone ever add anything to `install.tools` it will fail catastrophically with a difficult-to-debug error.

The correct solution, of course, is that `hack` should not be in anyone's path ever.

**Review by edsantiago — CHANGES_REQUESTED** [review]

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:187]: Slight preference for `>>` (append instead of clobber). I suspect there's nothing valuable in that file (if it even exists), but it seems better in principle.

Also, will any of this code ever run on a system with SELinux enabled? If so you may want to `chmod 600` and `restorecon` the file. Likewise with user file below.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:189]: 700 is bizarre, any particular reason for it?

File: Makefile
Comment (cevich) [reply Makefile:891]: > carries over for the lifetime of the make

Ugh, god, thanks for catching that.  Yeah, I really struggled with this fix and even trying to understand how/why it suddenly became a problem.  I tracked down the path modification to a change (facc009ca0) @vrothberg made at the end of April.  It appears related to a benchmarking change that I can't understand.

Valentin: The path change is causing a problem where `hack/install_bats.sh` is incorrectly finding `hack/bats`, so therefore failing to install when it's really needed.  Is there a simple change we can make so your benchmarks don't need `hack` in their path (since it's dangerous, as Ed pointed out).

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:187]: There should never be a case where anything is trying to ssh-out from the test VM to anything other than localhost (it would indicate a security problem AFAIC).  We *want* the absolute authority over the contents in this case, so overwriting was intentional.  I can add a comment to that effect if you think it would help.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:189]: It keeps the permissions consistent with all the files under `.ssh`.

File: Makefile
Comment (cevich) [reply Makefile]: While waiting for @vrothberg, I just force-pushed a commit that fixes the `Makefile`.  Lets see what breaks, if anything...

**Comment by cevich** [comment]
Force-push: Rebased + added comments re: ssh setup.

File: Makefile
Comment (vrothberg) [reply Makefile]: @cevich, feel free to revert this specific change.  I can export the path in `make localbenchmarks`.  I added it to have some scripts in ./hack in the PATH.

**Comment by cevich** [comment]
force push: rebased + fixed 'artifacts' task failure.

File: Makefile
Comment (cevich) [reply Makefile]: As Ed found out, probably the only safe way to do it in a recipe (that doesn't affect every other target) would be something like:

```Makefile
.PHONY: localbenchmarks
localbenchmarks:
    env PATH=$(PATH):$(CURDIR)/hack command-to-run
```

Just be aware, this can cause some really strange and hard to debug side-effects, like `command-to-run` picking up `hack/bats` instead of `/usr/bin/bats` (for example) :wink:

**Comment by cevich** [comment]
Oof...so...many...flakes :cry:

**Comment by cevich** [comment]
Force-push: Rebased + split `Makefile` fix into dedicated commit.

**Pull request merged**
